File: src/replay.cpp
#include "replay.hpp"
#include <iostream>

using namespace std;
using namespace rr;

ReplayingFlags replay_flags_to_rust(rr::ReplayFlags in) {
    ReplayingFlags out = {};
    out.dont_launch_debugger = in.dont_launch_debugger;
    out.goto_event = in.goto_event;
    out.singlestep_to_event = in.singlestep_to_event;
    out.target_process = in.target_process;
    out.dbg_port = in.dbg_port;
    out.dbg_host = rust::String(in.dbg_host);
    out.keep_listening = in.keep_listening;
    rust::Vec<rust::String> gdb_options;
    for (auto option : in.gdb_options){
        gdb_options.push_back(rust::String(option));
    }
    out.gdb_options = gdb_options;
    out.gdb_binary_file_path = rust::String(in.gdb_binary_file_path);
    out.redirect = in.redirect;
    out.cpu_unbound = in.cpu_unbound;
    out.share_private_mappings = in.share_private_mappings;
    out.dump_interval = in.dump_interval;
    out.serve_files = in.serve_files;
    out.tty = rust::String(in.tty);
    return out;
}
rr::ReplayFlags replay_flags_to_cpp(ReplayingFlags in){
    rr::ReplayFlags out;
    out.dont_launch_debugger = in.dont_launch_debugger;
    out.goto_event = in.goto_event;
    out.singlestep_to_event = in.singlestep_to_event;
    out.target_process = in.target_process;
    out.dbg_port = in.dbg_port;
    out.dbg_host = std::string(in.dbg_host);
    out.keep_listening = in.keep_listening;
    std::vector<std::string> gdb_options;
    for (auto option : in.gdb_options){
        gdb_options.push_back(std::string(option));
    }
    out.gdb_options = gdb_options;
    out.gdb_binary_file_path = std::string(in.gdb_binary_file_path);
    out.redirect = in.redirect;
    out.cpu_unbound = in.cpu_unbound;
    out.share_private_mappings = in.share_private_mappings;
    out.dump_interval = in.dump_interval;
    out.serve_files = in.serve_files;
    out.tty = std::string(in.tty);
    return out;
}
int replay(ReplayingFlags flags, rust::String tracedir){
    rr::ReplayFlags rrflags = replay_flags_to_cpp(flags);
    return rr::start_replaying(rrflags, std::string(tracedir));
}
ReplayingFlags get_default_replay_flags(){
    rr::ReplayFlags flags;
    return replay_flags_to_rust(flags);
}
ReplayingFlags replay_flags_pipe_test(ReplayingFlags flags){
    rr::ReplayFlags temp = replay_flags_to_cpp(flags);
    return replay_flags_to_rust(temp);
}


static ReplaySession::Flags session_flags(const rr::ReplayFlags& flags) {
  ReplaySession::Flags result;
  result.redirect_stdio = flags.redirect;
  result.redirect_stdio_file = flags.tty;
  result.share_private_mappings = flags.share_private_mappings;
  result.cpu_unbound = flags.cpu_unbound;
  return result;
}
namespace rr {
std::unique_ptr<ReplayController> new_replay_controller(rust::String trace_dir, ReplayingFlags rust_flags){
    rr::ReplayFlags flags = replay_flags_to_cpp(rust_flags);
    GdbServer::Target target;
    switch (flags.process_created_how) {
    case ReplayFlags::CREATED_EXEC:
      target.pid = flags.target_process;
      target.require_exec = true;
      break;
    case ReplayFlags::CREATED_FORK:
      target.pid = flags.target_process;
      target.require_exec = false;
      break;
    case ReplayFlags::CREATED_NONE:
      break;
    }
    target.event = flags.goto_event;
    ReplaySession::shr_ptr replay_session =
        ReplaySession::create(std::string(trace_dir), session_flags(flags));
    return std::unique_ptr<ReplayController>(new ReplayController(replay_session, target));
}

void ReplayController::print_test_controller() const{
    std::cout << "THEE CRITICAL BEATDOWN" << std::endl;
}
void ReplayController::test_run() {
  while(true) {
    ReplayResult result =
        timeline.replay_step_forward(RUN_CONTINUE);
    if (result.status == REPLAY_EXITED) {
        std::cout << "Debugger was not launched before end of trace" << std::endl;
      return;
    }
  }
}
void ReplayController::setup() {
  ReplayResult result;
  do {
    result =
        timeline.replay_step_forward(RUN_CONTINUE);
    if (result.status == REPLAY_EXITED) {
      // TODO LOG(info) << "Debugger was not launched before end of trace";
      this->is_replay_finished = true;
      return;
    }
  } while (!at_target(result));
  Task* t = timeline.current_session().current_task();
  FrameTime first_run_event = std::max(t->vm()->first_run_event(),
    t->thread_group()->first_run_event());
  if (first_run_event) {
    timeline.set_reverse_execution_barrier_event(first_run_event);
  }
}
bool ReplayController::can_continue_replay() const {
    return !(this->is_replay_finished);
}

GdbRegisterValue new_register_value() {
    GdbRegisterValue reg;
    memset(&reg, 0, sizeof(reg));
    reg.size = 9;
    return reg;
}

} // namespace rr
void printmyval(){
  std::cout << "FYCJ YA" << std::endl;
}

File: src/replay.rs
use cxx::UniquePtr;
use std::pin::Pin;

#[cxx::bridge]
pub mod replayffi {


    #[derive(Debug, PartialEq, Eq, Clone)]
    pub struct ReplayingFlags {
        goto_event : i64,
        singlestep_to_event : i64,
        target_process : i32,
        // TODO ZACK: process_created_how
        dont_launch_debugger : bool,
        dbg_port : i32,
        dbg_host : String,
        keep_listening : bool,
        gdb_options : Vec<String>,
        gdb_binary_file_path : String,
        redirect : bool,
        cpu_unbound : bool,
        share_private_mappings : bool,
        dump_interval : u32,
        serve_files : bool,
        tty : String,

    }
    unsafe extern "C++" {
        include!("librr-rs/src/replay.hpp");
        pub fn replay(flags : ReplayingFlags, tracedir : String) -> i32;
        pub fn get_default_replay_flags() -> ReplayingFlags;
        fn replay_flags_pipe_test(flags: ReplayingFlags) -> ReplayingFlags;
    }
    #[namespace = "rr" ]
    unsafe extern "C++" {
        include!("librr-rs/src/replay.hpp");
        type ReplayController;
        fn print_test_controller(&self);
        fn test_run(self : Pin<&mut ReplayController>);
        fn can_continue_replay(&self) -> bool;
        fn setup(self : Pin<&mut ReplayController>);
        fn new_replay_controller(trace_dir : String, flags:ReplayingFlags) -> UniquePtr<ReplayController>;
    }
}

pub use replayffi::*;

#[cfg(test)]
mod tests {
    use super::*;
    #[test]
    fn replay_flags_fuck(){
        let flags = get_default_replay_flags();
        let mut controller = new_replay_controller("".to_owned(), flags);
        controller.print_test_controller();
        controller.pin_mut().setup();
        println!("can continue replay: {}", controller.can_continue_replay());
    }
    #[test]
    fn replay_controller_setup(){
        let flags = get_default_replay_flags();
        let mut controller = new_replay_controller("".to_owned(), flags);
        controller.print_test_controller();
        controller.pin_mut().setup();
        assert!(controller.can_continue_replay())
    }
    #[test]
    fn replay_flags_defaults(){
        //test_replay();
        let flags = get_default_replay_flags();
        assert_eq!(flags.dont_launch_debugger, false);
        assert_eq!(flags.goto_event, 0);
        assert_eq!(flags.singlestep_to_event, 0);
        assert_eq!(flags.target_process,0);
        assert_eq!(flags.dbg_port, -1);
        assert_eq!(flags.dbg_host, "127.0.0.1");
        assert_eq!(flags.keep_listening,false);
        assert_eq!(flags.redirect,true);
        assert_eq!(flags.cpu_unbound, false);
        assert_eq!(flags.share_private_mappings, false);
        assert_eq!(flags.dump_interval, 0);
        assert_eq!(flags.serve_files, false);
    }
    #[test]
    fn replay_flags_pipe_test_1(){
        let mut flags = get_default_replay_flags();
        flags.dont_launch_debugger = true;
        flags.goto_event = -1;
        flags.singlestep_to_event = 100;
        flags.target_process = 1;
        flags.dbg_port = 10;
        flags.dbg_host = "The Lands Between".to_owned();
        flags.keep_listening = true;
        flags.gdb_options.push("Try".to_owned());
        flags.gdb_options.push("Attacking".to_owned());
        flags.gdb_binary_file_path = "/bin/elden_ring".to_owned();
        flags.redirect = false;
        flags.cpu_unbound = true;
        flags.share_private_mappings = true;
        flags.dump_interval = 10;
        flags.serve_files = true;
        flags.tty = "Torrent".to_owned(); 
        let response = replay_flags_pipe_test(flags.clone());
        assert_eq!(flags, response);
    }

}
